src/types/opt.rs

```diff
@@ -0,0 +1,95 @@
+use std::fmt;
+
+use crate::NativeRepr;
+
+/// A convenience type to explicitly mark
+/// a function argument as `opt T`.
+///
+/// When left unspecified on Redscript side,
+/// it translates to its `Default` representation.
+#[derive(Default, Debug, Clone, PartialEq, PartialOrd)]
+pub enum Opt<T: NativeRepr> {
+    /// Value is specified and guaranteed to be non-`Default` value.
+    NonDefault(T),
+    /// `Default` value.
+    #[default]
+    Default,
+}
+
+impl<T> Opt<T>
+where
+    T: NativeRepr,
+{
+    pub fn into_option(self) -> Option<T> {
+        match self {
+            Self::NonDefault(x) => Some(x),
+            Self::Default => None,
+        }
+    }
+}
+
+impl<T> Opt<T>
+where
+    T: NativeRepr + Default,
+{
+    pub fn unwrap_or_default(self) -> T {
+        match self {
+            Self::NonDefault(x) => x,
+            Self::Default => T::default(),
+        }
+    }
+}
+
+impl<T> fmt::Display for Opt<T>
+where
+    T: fmt::Display + NativeRepr,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::NonDefault(x) => write!(f, "{}", x),
+            Self::Default => write!(f, "{}", <Self as Default>::default()),
+        }
+    }
+}
+
+impl<T> Copy for Opt<T> where T: NativeRepr + Copy {}
+impl<T> Eq for Opt<T> where T: NativeRepr + Eq {}
```

Comment:
seems these should be derived

Reply:
The thing is that I don't want to force `T` to impl `Copy` or `Eq` (which many types aren't), I just want to impl it if `T` already does. Otherwise people won't be able to use e.g. `Opt<f32>`.

Reply:
that's what the derive does already

Reply:
my bad! done in f01d044